### ISSUE TITLE
FTBFS: tests failures on [armhf,mips,mips64el,mipsel]

### ISSUE THREAD

**Issue by onlyjob** [Yelp/dumb-init#107]
1.1.2 currently [FTBFS](https://buildd.debian.org/status/package.php?p=dumb-init) on

 * [armhf](https://buildd.debian.org/status/fetch.php?pkg=dumb-init&arch=armhf&ver=1.1.2-1&stamp=1469941024)
 * [mips](https://buildd.debian.org/status/fetch.php?pkg=dumb-init&arch=mips&ver=1.1.2-1&stamp=1469941155)
 * [mips64el](https://buildd.debian.org/status/fetch.php?pkg=dumb-init&arch=mips64el&ver=1.1.2-1&stamp=1469940891)
 * [mipsel](https://buildd.debian.org/status/fetch.php?pkg=dumb-init&arch=mipsel&ver=1.1.2-1&stamp=1469941005)

(and some other architectures) due to test failures.

Interestingly 1.0.3 used to built successfully on those architectures...

**Comment by asottile** [comment]
I believe this is a duplicate of #106

**Comment by chriskuehl** [comment]
Also #105. I'll get these fixed, thanks for the report!

**Comment by chriskuehl** [comment]
I think these should be fixed except for the kfreebsd builds, where dumb-init isn't working at all.

I'll try to debug kfreebsd soon. It's not super surprising as we've never tested or tried to support anything but the Linux kernel, but I'm certainly open to supporting it if we can.

@onlyjob can you check if this fixed the build for those other architectures? I don't have an easy way to test on those platforms.

**Comment by onlyjob** [comment]
Thanks. :)
Unfortunately I have no access to most of those architectures. I'll upload new release (once it is ready) and we'll see how it goes from build logs. I should have an old mips VM somewhere but right now I have no time to revive it... Sorry.

**Comment by chriskuehl** [comment]
@onlyjob I've released 1.1.3 which should have test fixes for those platforms, plus working support for FreeBSD (tested on the kfreebsd jessie port, anyway).

**Comment by onlyjob** [comment]
1.1.3 now FTBFS on _x86_64_ (a.k.a. _amd64_) in _pbuilder_:

~~~~
tests/child_processes_test.py FFFF.................... 
tests/cli_test.py ...................................................................... 
tests/exit_status_test.py ................................................ 
tests/proxies_signals_test.py FFFFFFFFFFFFFFFFFFFFFFFFFF 
tests/shell_background_test.py FFFF 
tests/tty_test.py . 
~~~~

Many processes like `dumb-init /usr/bin/python -m testing.print_signals` (and others) are left behind:

~~~~
        ├─4*[dumb-init───sh─┬─2*[tail]]
        │                   └─2*[yes]]
        ├─30*[dumb-init───python]
~~~~

I did not look deep into the problem but many `tail` processes are actively using CPU so clearly something unexpected is happening...

**Comment by chriskuehl** [comment]
That is really weird, this has passed several times on our own CI (Circle + internal Jenkins) and locally. All of this is amd64.

I'll look into this, thanks.

**Comment by chriskuehl** [comment]
Ah, I was finally able to reproduce this. The good news is that the binary itself seems to be fine (so we didn't make a broken release or anything), but the tests fail on Debian stretch for me with an error like this:

```
    def child_pids(pid):
        """Return a list of direct child PIDs for the given PID."""
        children = set()
        for p in LocalPath('/proc').listdir():
            stat = p.join('stat')
            if stat.isfile():
                stat = stat.open().read()
                m = re.match('^\d+ \([^\)]+\) [a-zA-Z] (\d+) ', stat)
>               assert m, stat
E               AssertionError: 2127 ((sd-pam)) S 2125 2125 2125 0 -1 1077936448 23 0 0 0 0 0 0 0 20 0 1 0 3443 80334848 476 18446744073709551615 1 1 0 0 0 0 0 4096 0 0 0 0 17 3 0 0 0 0 0 0 0 0 0 0 0 0 0

testing/__init__.py:57: AssertionError
```

This is because we changed the tests to stop using Linux-specific `/proc` features so that the tests would pass on FreeBSD. Looks like the way we're parsing the `/proc/*/stat` files is not totally correct and fails in some cases, though. Will fix this, thanks!

**Comment by chriskuehl** [comment]
@onlyjob the fix is merged, can you see if the build succeeds now? I confirmed it does on stretch for me.

Since this is a test-only change we don't plan to issue a new release.

**Comment by chriskuehl** [comment]
Looks like 1.2.0 has built successfully on these architectures: https://buildd.debian.org/status/package.php?p=dumb-init&suite=stretch

Thanks again @onlyjob!
